### PR TITLE
CircleCI: Enforce strict TypeScript typechecking for Gutenboarding

### PR DESCRIPTION
**Pull request by ockham**
#### Changes proposed in this Pull Request

CircleCI: Enforce strict TypeScript typechecking for Gutenboarding

#### Testing instructions

CircleCI should be all-green, especially for the `typecheck-strict` task.

/cc @sirreal

### CODE REVIEW

File: .circleci/config.yml
Comment (sirreal) [inline .circleci/config.yml]: We could call it something more descriptive: `typecheck-gutenboarding`?

Comment (ockham) [reply .circleci/config.yml]: I was considering that, but then decided that we might not make it project-specific but might later add other projects to strictly check here -- vs giving them all a separate CircleCI targets. (My choice is reflected by the [`name`](https://github.com/Automattic/wp-calypso/pull/37945/files#diff-1d37e48f9ceff6d8030570cd36286a61R224), `TypeScript strict typecheck of individual subprojects`)

I don't feel super strong about this -- would you prefer a project-specific target name?

Comment (blowery) [reply .circleci/config.yml]: +1 for keeping it generic, but a soft +1. Either way is fine, this is pretty easy to change later.

Comment (sirreal) [reply .circleci/config.yml]: Generic seems fine, it's easy enough to change 👍

**Comment by sirreal** [comment]
#38401 should unblock this 🎉

File: .circleci/config.yml
Comment (simison) [inline .circleci/config.yml]: Keeping to the generic spirit, there might soon be other folders listed here. Once that happens, might make sense to pull them from a file, or keep the list in `package.json` for example by adding `npm run typecheck-strict` or similar?

No action needed for this PR. :-)

**Review by simison — APPROVED** [review]
Not an expert here but always supportive of things like this. 👍

**Comment by simison** [comment]
Completely ignorant question since I'm not super familiar with this stuff — not blocking merging: I know type checking is initially slow on my computer and faster on sequential rounds. Did you check if this slows CI at all significantly and if it's taking advantage of caching?

**Comment by sirreal** [comment]
> I know type checking is initially slow on my computer and faster on sequential rounds. Did you check if this slows CI at all significantly and if it's taking advantage of caching?

Anecdotally: The most recent run of the new job took 37s, most of which was attaching the workspace. The type checking itself ran in 7s.

There's no caching involved at this time, I don't think it's necessary and speculate it would be marginal improvement, if any.

**Comment by ockham** [comment]
> Completely ignorant question since I'm not super familiar with this stuff — not blocking merging: I know type checking is initially slow on my computer and faster on sequential rounds. Did you check if this slows CI at all significantly and if it's taking advantage of caching?

I didn't, TBH. Are you talking about type checks in the editor (which take a long time, since tsserver has to parse all of Calypso), or the command line? Running TS on _just_ Gutenboarding (`npm run typecheck -- --project client/landing/gutenboarding`) seems fairly quick for me (and that's what CircleCI does here); I couldn't find any resources on whether that's cached at all.

**Comment by simison** [comment]
>Are you talking about type checks in the editor (which take a long time, since tsserver has to parse all of Calypso), or the command line? 

Editor 👍 — didn't test run the CLI command locally; figured CI is already doing it for us. :-)

**Pull request merged**
